packages/devtools/lib/src/timeline/event_details.dart

```diff
@@ -266,71 +321,116 @@ class _UiEventDetails extends CoreElement {
+
+    canvas.onStackFrameSelected.listen((CpuStackFrame stackFrame) {
+      stackFrameDetails.text = stackFrame.toString();
     });
 
-    flameChart.add(flameChartCanvas.element);
+    add(canvas.element);
   }
 
-  void _updateFlameChartForError(CoreElement errorDiv) {
-    flameChart.add(errorDiv);
-    showingFlameChartError = true;
+  void _updateChartWithError(String message) {
+    showingError = true;
+    error.text = message;
+    error.attribute('hidden', false);
     stackFrameDetails.attribute('hidden', true);
   }
 
   Future<void> update(TimelineEvent event) async {
-    if (event == this.event) {
-      return;
-    }
-
     reset();
     this.event = event;
 
-    final Spinner spinner = Spinner()..clazz('cpu-profile-spinner');
-    add(spinner);
+    // Update the canvas if the flame chart is visible. Otherwise, mark the
+    // canvas as needing a rebuild.
+    if (!isHidden) {
+      final Spinner spinner = Spinner()..clazz('cpu-profile-spinner');
+      add(spinner);
+
+      try {
+        await _drawFlameChart();
+
+        if (!showingError) {
+          stackFrameDetails.text = stackFrameDetailsDefaultText;
+          stackFrameDetails.attribute('hidden', false);
+        }
+      } catch (e) {
+        _updateChartWithError('Error retrieving CPU profile: ${e.toString()}');
+      }
 
-    try {
-      // TODO(kenzie): add a timeout here so we don't appear to have an
-      // infinite spinner.
-      await _drawFlameChart();
-    } catch (e) {
-      _updateFlameChartForError(div(
-          text: 'Error retrieving CPU profile: ${e.toString()}', c: 'message'));
+      spinner.element.remove();
+    } else {
+      canvasNeedsRebuild = true;
     }
+  }
 
-    spinner.element.remove();
+  void updateForContainerResize() {
+    if (canvas == null) {
+      return;
+    }
 
-    stackFrameDetails.text = stackFrameDetailsDefaultText;
+    // Only update the canvas if the flame chart is visible and has data.
+    // Otherwise, mark the canvas as needing a rebuild.
+    if (!isHidden && cpuProfileData != null) {
+      canvas.forceRebuildForSize(
+        canvas.flameChartWidthWithInsets,
+        math.max(
+          // Subtract [rowHeightWithPadding] to account for size of
+          // [stackFrameDetails] section at bottom of chart.
+          element.scrollHeight.toDouble() - rowHeightWithPadding,
+          // Add 1 to account for a row of padding at bottom of chart.
+          (cpuProfileData.cpuProfileRoot.depth + 1) * rowHeightWithPadding,
+        ),
+      );
+    } else {
+      canvasNeedsRebuild = true;
+    }
   }
 
-  void reset() {
-    flameChart.clear();
-    stackFrameDetails.clear();
-    showingFlameChartError = false;
-    stackFrameDetails.attribute(
-        'hidden', selectedTab != EventDetailsTabType.flameChart);
-    cpuProfileData = null;
+  Future<void> show() async {
+    attribute('hidden', false);
+
+    if (canvasNeedsRebuild) {
+      assert(event != null);
+      await update(event);
```

Comment:
should 
`canvasNeedsRebuild = false`
 be set before awaiting the call to update?

Reply:
I don't think it matters because nothing in the update flow depends on `canvasNeedsRebuild`. I can move it, though, in case something in the future does.